<a href="https://colab.research.google.com/github/TReV-89/TReV-89/blob/main/LanguuageID.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

LANGUAGE MODEL USING COUNTVECTORTIZER FUNCTION AND PIPELINE METHOD WITH N-GRAMS

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.naive_bayes import MultinomialNB, ComplementNB, BernoulliNB
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import classification_report,accuracy_score

In [2]:
df = pd.read_json("/content/drive/MyDrive/salt-train-v1.2.jsonl", lines = True)
df.pop("tts-speech")
data = {
    "language": [],
    "text": []
}
for row in df['text']:
    for key, value in row.items():
        data["language"].append(key)
        data["text"].append(value)
df1 = pd.DataFrame(data,columns = ['language','text'])
df1.head()

language                                               text
0      eng                   It was not a ghost refugee camp.
1      lug  Enkambi y'abanoonyiboobubudamu teyaliiwo mu bu...
2      ach                             Pe obedo kem goba goba
3      teo                      Mam arai ekabi lo erai ekwam.
4      lgg  Eri aa'ni ndra kembe emunyale eyini aa'zu inzo...

In [3]:
X = df1['text']
Y = df1['language']
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.33, random_state=42)


In [4]:
count = CountVectorizer(analyzer = 'char',ngram_range=(2,6))

pipeline_MultinominalNB = Pipeline([
   ('vectorizer',count),
   ('model',MultinomialNB())
])
pipeline_MultinominalNB.fit(X_train,Y_train)
Y_pred = pipeline_MultinominalNB.predict(X_test)

print(classification_report(Y_test,Y_pred))

accuracy = accuracy_score(Y_test, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00      7826
         eng       1.00      1.00      1.00      7783
         lgg       1.00      1.00      1.00      8089
         lug       1.00      1.00      1.00      7799
         nyn       1.00      1.00      1.00      7967
         teo       1.00      1.00      1.00      7952

    accuracy                           1.00     47416
   macro avg       1.00      1.00      1.00     47416
weighted avg       1.00      1.00      1.00     47416

Accuracy: 0.9992407626117766


In [72]:
count = CountVectorizer(analyzer = 'char',ngram_range=(2,3))

pipeline_BernoulliNB = Pipeline([
   ('vectorizer',count),
   ('model',BernoulliNB())
])
pipeline_BernoulliNB.fit(X_train,Y_train)
Y_pred = pipeline_BernoulliNB.predict(X_test)

print(classification_report(Y_test,Y_pred))

accuracy = accuracy_score(Y_test, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00      7826
         eng       1.00      1.00      1.00      7783
         lgg       1.00      1.00      1.00      8089
         lug       0.99      1.00      0.99      7799
         nyn       1.00      0.99      1.00      7967
         teo       1.00      1.00      1.00      7952

    accuracy                           1.00     47416
   macro avg       1.00      1.00      1.00     47416
weighted avg       1.00      1.00      1.00     47416

Accuracy: 0.9970895900118104


In [57]:
count = CountVectorizer(analyzer = 'char',ngram_range=(2,7))

pipeline_ComplementNB = Pipeline([
   ('vectorizer',count),
   ('model',ComplementNB())
])
pipeline_ComplementNB.fit(X_train,Y_train)
Y_pred = pipeline_ComplementNB.predict(X_test)

print(classification_report(Y_test,Y_pred))

accuracy = accuracy_score(Y_test, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00      7826
         eng       1.00      1.00      1.00      7783
         lgg       1.00      1.00      1.00      8089
         lug       1.00      1.00      1.00      7799
         nyn       1.00      1.00      1.00      7967
         teo       1.00      1.00      1.00      7952

    accuracy                           1.00     47416
   macro avg       1.00      1.00      1.00     47416
weighted avg       1.00      1.00      1.00     47416

Accuracy: 0.999156402901974


Multinomial showed highest accuracy out of the three different Naive Bayes models of 99.7%

In [7]:
df2 = pd.read_json("/content/drive/MyDrive/salt-test-v1.2.jsonl", lines = True)
df2.pop("tts-speech")
data1 = {
    "language": [],
    "text": []
}
for row in df2['text']:
    for key, value in row.items():
        data1["language"].append(key)
        data1["text"].append(value)
df3 = pd.DataFrame(data1,columns = ['language','text'])
df3.head()

language                                               text
0      eng  The fashion industry is starting to thrive again.
1      lug      Ekisaawe ky'emisono kitandise okusituka nate.
2      ach  Yub me cital ruk mapatpat manyen tye ka dongo ...
3      teo            Ageutu ikampunin luka enape apolo bobo.
4      lgg   Okalamvu suta o'diru 'diyini 'diyi 'ye e'do tutu

In [44]:
X = df3['text']
Y = df3['language']

Y_pred = pipeline_MultinominalNB.predict(X)

print(classification_report(Y,Y_pred))

accuracy = accuracy_score(Y, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00       500
         eng       1.00      1.00      1.00       500
         lgg       1.00      1.00      1.00       500
         lug       1.00      1.00      1.00       500
         nyn       1.00      1.00      1.00       500
         teo       1.00      1.00      1.00       500

    accuracy                           1.00      3000
   macro avg       1.00      1.00      1.00      3000
weighted avg       1.00      1.00      1.00      3000

Accuracy: 0.9986666666666667


In [59]:
Y_pred = pipeline_ComplementNB.predict(X)

print(classification_report(Y,Y_pred))

accuracy = accuracy_score(Y, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00       500
         eng       1.00      1.00      1.00       500
         lgg       1.00      1.00      1.00       500
         lug       1.00      1.00      1.00       500
         nyn       1.00      1.00      1.00       500
         teo       1.00      1.00      1.00       500

    accuracy                           1.00      3000
   macro avg       1.00      1.00      1.00      3000
weighted avg       1.00      1.00      1.00      3000

Accuracy: 0.9986666666666667


In [64]:
Y_pred = pipeline_BernoulliNB.predict(X)

print(classification_report(Y,Y_pred))

accuracy = accuracy_score(Y, Y_pred)
print("Accuracy:", accuracy)

              precision    recall  f1-score   support

         ach       1.00      1.00      1.00       500
         eng       1.00      1.00      1.00       500
         lgg       1.00      1.00      1.00       500
         lug       1.00      1.00      1.00       500
         nyn       1.00      1.00      1.00       500
         teo       1.00      1.00      1.00       500

    accuracy                           1.00      3000
   macro avg       1.00      1.00      1.00      3000
weighted avg       1.00      1.00      1.00      3000

Accuracy: 0.9986666666666667


Bernoulli has a higher accuracy on the test SALT dataset

Use Complent Bernoulli and Multinomial for test samaples and inference code



In [30]:
new_samples = ['I am going to the katale', # eng
'Ngenda mu katale', # lug
'Uganda', # any
'Ekikopo', # lug or nyn
'Runyankole', # nyn or lug
'Omuntu', # nyn or lug
"Ebigambo by'amagezi", # lug
"Bw'okoowa", # lug
"yiga", # lug
"kubivaako", # lug
"Mulindeeko tujja kubayita", # lug
"Gamba Mpurire", # nyn
"Ekiro twaburayo kimwe kyonka tukajaguza embaga y'omwaka", # nyn
"Tushangye", # nyn
"Ekitongore ekya UWA kiriyo nikibangura abanyamakuru kuruga omukyanga ekya Kigezi ahakuhandiika amakuru agakwatiraine n’okurinda enyamaishwa z’omukishaka.", # nyn
"itye nining", # ach
"Irii maber, ladit?", # ach
"Ibuto maber", # ach
"Ibuto Gulu", # ach,
"Amito tedo labolo kwon.", # ach
"Ngo manyen?", # ach
"Wacito I gang kwan nino ducu", # ach
"Kumbedi abedo Anaka i Amuru aa ki California i Amerika ento." # ach
]

FOR THE MULTINOMIALNB MODEL

In [60]:
y_pred_new = pipeline_MultinominalNB.predict(new_samples)

for sentence, language in zip(new_samples, y_pred_new):
    print(f"Text: '{sentence}' => Predicted Language: {language}")

Text: 'I am going to the katale' => Predicted Language: eng
Text: 'Ngenda mu katale' => Predicted Language: lug
Text: 'Uganda' => Predicted Language: lug
Text: 'Ekikopo' => Predicted Language: nyn
Text: 'Runyankole' => Predicted Language: lug
Text: 'Omuntu' => Predicted Language: nyn
Text: 'Ebigambo by'amagezi' => Predicted Language: lug
Text: 'Bw'okoowa' => Predicted Language: lug
Text: 'yiga' => Predicted Language: lug
Text: 'kubivaako' => Predicted Language: lug
Text: 'Mulindeeko tujja kubayita' => Predicted Language: lug
Text: 'Gamba Mpurire' => Predicted Language: nyn
Text: 'Ekiro twaburayo kimwe kyonka tukajaguza embaga y'omwaka' => Predicted Language: nyn
Text: 'Tushangye' => Predicted Language: nyn
Text: 'Ekitongore ekya UWA kiriyo nikibangura abanyamakuru kuruga omukyanga ekya Kigezi ahakuhandiika amakuru agakwatiraine n’okurinda enyamaishwa z’omukishaka.' => Predicted Language: nyn
Text: 'itye nining' => Predicted Language: ach
Text: 'Irii maber, ladit?' => Predicted Language

In [73]:
y_pred_new = pipeline_BernoulliNB.predict(new_samples)

for sentence, language in zip(new_samples, y_pred_new):
    print(f"Text: '{sentence}' => Predicted Language: {language}")

Text: 'I am going to the katale' => Predicted Language: eng
Text: 'Ngenda mu katale' => Predicted Language: lug
Text: 'Uganda' => Predicted Language: ach
Text: 'Ekikopo' => Predicted Language: ach
Text: 'Runyankole' => Predicted Language: lug
Text: 'Omuntu' => Predicted Language: lug
Text: 'Ebigambo by'amagezi' => Predicted Language: lug
Text: 'Bw'okoowa' => Predicted Language: lug
Text: 'yiga' => Predicted Language: ach
Text: 'kubivaako' => Predicted Language: lug
Text: 'Mulindeeko tujja kubayita' => Predicted Language: lug
Text: 'Gamba Mpurire' => Predicted Language: nyn
Text: 'Ekiro twaburayo kimwe kyonka tukajaguza embaga y'omwaka' => Predicted Language: nyn
Text: 'Tushangye' => Predicted Language: nyn
Text: 'Ekitongore ekya UWA kiriyo nikibangura abanyamakuru kuruga omukyanga ekya Kigezi ahakuhandiika amakuru agakwatiraine n’okurinda enyamaishwa z’omukishaka.' => Predicted Language: nyn
Text: 'itye nining' => Predicted Language: ach
Text: 'Irii maber, ladit?' => Predicted Language

In [62]:
y_pred_new = pipeline_ComplementNB.predict(new_samples)

for sentence, language in zip(new_samples, y_pred_new):
    print(f"Text: '{sentence}' => Predicted Language: {language}")

Text: 'I am going to the katale' => Predicted Language: eng
Text: 'Ngenda mu katale' => Predicted Language: lug
Text: 'Uganda' => Predicted Language: lug
Text: 'Ekikopo' => Predicted Language: lug
Text: 'Runyankole' => Predicted Language: lug
Text: 'Omuntu' => Predicted Language: nyn
Text: 'Ebigambo by'amagezi' => Predicted Language: lug
Text: 'Bw'okoowa' => Predicted Language: lug
Text: 'yiga' => Predicted Language: lug
Text: 'kubivaako' => Predicted Language: lug
Text: 'Mulindeeko tujja kubayita' => Predicted Language: lug
Text: 'Gamba Mpurire' => Predicted Language: nyn
Text: 'Ekiro twaburayo kimwe kyonka tukajaguza embaga y'omwaka' => Predicted Language: nyn
Text: 'Tushangye' => Predicted Language: nyn
Text: 'Ekitongore ekya UWA kiriyo nikibangura abanyamakuru kuruga omukyanga ekya Kigezi ahakuhandiika amakuru agakwatiraine n’okurinda enyamaishwa z’omukishaka.' => Predicted Language: nyn
Text: 'itye nining' => Predicted Language: ach
Text: 'Irii maber, ladit?' => Predicted Language